xcube/core/store/preload.py

```diff
@@ -0,0 +1,409 @@
+            if event.exception is not None:
+                self.exception = event.exception
+
+    def __str__(self):
+        return ", ".join(f"{k}={v}" for k, v in _to_dict(self).items())
+
+    def __repr__(self):
+        args = ", ".join(f"{k}={v!r}" for k, v in _to_dict(self).items())
+        return f"{self.__class__.__name__}({args})"
+
+
+class PreloadHandle(ABC):
+    """A handle for a preload job.
+
+    TODO: add more doc
+    """
+
+    @abstractmethod
+    def get_state(self, data_id: str) -> PreloadState:
+        """Get the preload state for the given *data_id*.
+
+        Args:
+            data_id: The data identifier.
+        Returns:
+            The preload state.
+        """
+
+    @property
+    @abstractmethod
+    def cancelled(self) -> bool:
+        """True` if the preload job has been cancelled."""
+
+    @abstractmethod
+    def cancel(self):
+        """Cancel the preload job."""
+
+    def close(self):
+        """Close the preload job.
+
+        Should be called if the preloaded data is no longer needed.
+
+        This method usually cleans the cache associated with
+        this preload object.
+
+        The default implementation does nothing.
+        """
+
+    @abstractmethod
+    def show(self) -> Any:
+        """Show the current state of the preload job.
+
+        This method is useful for non-blocking / asynchronous preload
+        implementations, especially in a Jupyter Notebook context.
+        In this case an implementation might want to display a widget
+        suitable for in-place updating, e.g., an ``ipywidgets`` widget.
+        """
+
+    def notify(self, event: PreloadState):
+        """Notify about a preload state change.
+
+        Updates the preload job using the given partial state
+        *event* that refers to state changes of a running
+        preload task.
+
+        Args:
+            event: A partial state
+        """
+
+    def __enter__(self) -> "PreloadHandle":
+        """Enter the context.
+
+        Returns:
+            This object.
+        """
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit the context. Calls ``close()``."""
+        self.close()
+
+
+class NullPreloadHandle(PreloadHandle):
+    """Null-pattern implementation of a ``PreloadHandle``."""
+
+    def get_state(self, data_id: str) -> PreloadState:
+        return PreloadState(data_id)
+
+    @property
+    def cancelled(self) -> bool:
+        return False
+
+    def cancel(self):
+        pass
+
+    def show(self) -> Any:
+        return None
+
+
+class ExecutorPreloadHandle(PreloadHandle):
+    """TODO - Add docs"""
+
+    def __init__(
+        self,
+        data_ids: tuple[str, ...],
+        preload_data: Callable[[PreloadHandle, str], None] | None = None,
+        executor: Executor | None = None,
+        blocking: bool = True,
+    ):
+        self._preload_data = preload_data
+        self._executor = executor or ThreadPoolExecutor()
```

Comment:
I also wanted to discuss the number of threads that can be started concurrently which would vary according to the different machines that the users would use. We should maybe try to figure out what is the maximum number of threads to start also based on the tasks in the preload_data() function. Should this be done in the preload_data where the user defines the executor or we take care of that in the xcube interface, that is up for discussion.

Reply:
> Should this be done in the preload_data where the user defines the executor

Yes, clients can add their own `executor`.

> or we take care of that in the xcube interface

No.